### ISSUE TITLE
CatchupSubscriptionModel does not persist position and stops working for new events without 'hostinfo'

### ISSUE THREAD

**Issue by gavvvr** [johanhaleby/occurrent#142]
Hi

This one is easy to reproduce. Just create a free shared instance of MongoDB in Atlas, it does not allow 'hostinfo'.
After that run the following code 3 times to reproduce 2 issues from the title (I guess all of them are caused by uncaught NPE).
Here is the code:

```kotlin
val converter = JacksonCloudEventConverter<SomeSpecificDomainEvent>(jacksonObjectMapper(), URI("urn:test"))

data class SomeSpecificDomainEvent(val id: Int)


fun main() {
    val connectionString =
        "mongodb+srv://?:?@occurrent-test.g1kojdq.mongodb.net/?retryWrites=true&w=majority"
    val mongoClient = MongoClients.create(connectionString)
    val dbName = "occurrent-test"
    val database = mongoClient.getDatabase(dbName)

    val eventsCollectionName = "events"
    val eventStore =
        MongoEventStore(mongoClient, dbName, eventsCollectionName, EventStoreConfig(TimeRepresentation.RFC_3339_STRING))

    val nativeMongoSubscriptionModel = NativeMongoSubscriptionModel(
        database,
        eventsCollectionName,
        TimeRepresentation.RFC_3339_STRING,
        Executors.newCachedThreadPool(),
        RetryStrategy.retry().maxAttempts(1),
    )
    val positionStorage =
        NativeMongoSubscriptionPositionStorage(database, "subscription-position-collection")
    val wrappedSubscriptionModel = DurableSubscriptionModel(
        nativeMongoSubscriptionModel,
        positionStorage,
    )
    val catchupSubscriptionModel = CatchupSubscriptionModel(
        wrappedSubscriptionModel,
        eventStore,
        CatchupSubscriptionModelConfig(useSubscriptionPositionStorage(positionStorage))
    )

    // Now subscribe!
    catchupSubscriptionModel.subscribe("logger-subscription") { e ->
        println("Event number=${converter.toDomainEvent(e).id} received!")
    }

    when (eventStore.count()) {
        0L -> {
            println("Writing very first event to db")
            eventStore.write("some-stream", converter.toCloudEvent(SomeSpecificDomainEvent(1)))
        }

        1L -> {
            println("Writing very 2nd event to db")
            eventStore.write("some-stream", converter.toCloudEvent(SomeSpecificDomainEvent(2)))
        }
    }
}
```

On first (and each) run you will see a WARNing about 'hostinfo' and an NPE following soon after that:
```
[main] WARN  o.o.s.m.n.b.NativeMongoSubscriptionModel - Failed to get global subscription position from MongoDB, probably because the server doesn't allow to execute the "hostinfo" command. This only affects the very first event received by the subscription. If the processing of this event fails _and_ the application is restarted the event cannot be retried. If this is major concern, consider upgrading your MongoDB server to a non-shared environment that supports the "hostinfo" command. Error is:
Command failed with error 8000 (AtlasError): 'CMD_NOT_ALLOWED: hostInfo' on server ac-udi6xr8-shard-00-01.g1kojdq.mongodb.net:27017. The full response is {"ok": 0, "errmsg": "CMD_NOT_ALLOWED: hostInfo", "code": 8000, "codeName": "AtlasError"}
Exception in thread "pool-1-thread-1" java.lang.NullPointerException: SubscriptionPosition cannot be null
	at java.base/java.util.Objects.requireNonNull(Objects.java:235)
	at org.occurrent.subscription.StartAt$StartAtSubscriptionPosition.<init>(StartAt.java:99)
	at org.occurrent.subscription.StartAt.subscriptionPosition(StartAt.java:134)
	at org.occurrent.subscription.blocking.durable.catchup.CatchupSubscriptionModel.lambda$subscribe$7(CatchupSubscriptionModel.java:186)
	at org.occurrent.subscription.StartAt$Dynamic.get(StartAt.java:85)
	at org.occurrent.subscription.mongodb.nativedriver.blocking.NativeMongoSubscriptionModel.newInternalSubscription(NativeMongoSubscriptionModel.java:172)
	at org.occurrent.subscription.mongodb.nativedriver.blocking.NativeMongoSubscriptionModel.lambda$subscribe$1(NativeMongoSubscriptionModel.java:156)
	at org.occurrent.retry.internal.RetryExecution.lambda$executeWithRetry$2(RetryExecution.java:75)
	at org.occurrent.retry.internal.RetryExecution.lambda$executeWithRetry$4(RetryExecution.java:82)
	at org.occurrent.retry.internal.RetryExecution.lambda$executeWithRetry$3(RetryExecution.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
```

If you run the code for the 2nd time, you will see:
> Event number=1 received!
> Writing very 2nd event to db

and the NPE again. You would expect "Event number=2 received!" too, but nothing happens. So it means that after the NPE the subscription model is broken.

Run the above code for the 3rd time and you will see:
> **Event number=1 received!**
> Event number=2 received!

Receiving 1st event for the 2nd time means that subscription's position was not saved and it's not durable.
Any subsequent execution of the above code will lead to processing all events from the very beginning as if the subscription is not durable.

**Comment by johanhaleby** [comment]
Hi,

Thanks for the very detailed description. I'm running applications in Atlas free tier myself with Occurrent, and I remember that I had problems with it in the past. Have you tried "Restart Subscription when Oplog Lost" (search for it in the docs) or add this to subscription model:

```java
var subscriptionModel = new SpringMongoSubscriptionModel(mongoTemplate, SpringSubscriptionModelConfig.withConfig("events", TimeRepresentation.RFC_3339_STRING).restartSubscriptionsOnChangeStreamHistoryLost(true));
```

What's important is `restartSubscriptionsOnChangeStreamHistoryLost(true))`.

**Comment by gavvvr** [comment]
Hi @johanhaleby 
Thank you for fast response!

My original app is non-Spring based and there is no `restartSubscriptionsOnChangeStreamHistoryLost` property on `NativeMongoSubscriptionModel`.

Anyway, this does not help me even with Spring. The NPE and both problems still exists (position is not saved and no new events processed)
Here is my Spring setup with `restartSubscriptionsOnChangeStreamHistoryLost(true)`:
```kotlin
val converter = JacksonCloudEventConverter<SomeSpecificDomainEvent>(jacksonObjectMapper(), URI("urn:test"))
data class SomeSpecificDomainEvent(val id: Long)

@SpringBootApplication
class OccurrentSpringMongoApplication {
    @Bean
    fun eventStore(mongoTemplate: MongoTemplate, dbFactory: MongoDatabaseFactory): SpringMongoEventStore {
        val eventStoreConfig =
            EventStoreConfig.Builder()
                .eventStoreCollectionName("events")
                .transactionConfig(MongoTransactionManager(dbFactory))
                .timeRepresentation(TimeRepresentation.RFC_3339_STRING)
                .build()
        return SpringMongoEventStore(mongoTemplate, eventStoreConfig)
    }

    @Bean
    fun subscriptionPositionStorage(mongoTemplate: MongoTemplate): SubscriptionPositionStorage =
        SpringMongoSubscriptionPositionStorage(mongoTemplate, "subscriptions")

    @Bean
    fun subscriptionModel(
        storage: SubscriptionPositionStorage,
        mongoTemplate: MongoTemplate,
        eventStoreQueries: EventStoreQueries
    ): SubscriptionModel {
        val config = SpringMongoSubscriptionModelConfig.withConfig("events", TimeRepresentation.RFC_3339_STRING).restartSubscriptionsOnChangeStreamHistoryLost(true)
        val mongoSubscriptionModel =
            SpringMongoSubscriptionModel(mongoTemplate, config)
        val durableSubscriptionModel = DurableSubscriptionModel(mongoSubscriptionModel, storage)
        return CatchupSubscriptionModel(
            durableSubscriptionModel, eventStoreQueries,
            CatchupSubscriptionModelConfig(SubscriptionPositionStorageConfig.useSubscriptionPositionStorage(storage))
        )
    }
}

@Component
class Actions(val subscriptionModel: SubscriptionModel, val eventStore: SpringMongoEventStore) {
    @PostConstruct
    fun doWork() {
        val currentEventCount = eventStore.count()
        when (currentEventCount) {
            0L -> {
                println("Writing very first event to db")
                eventStore.write("some-stream", converter.toCloudEvent(SomeSpecificDomainEvent(1)))
            }

            1L -> {
                println("Writing 2nd event to db")
                eventStore.write("some-stream", converter.toCloudEvent(SomeSpecificDomainEvent(2)))
            }

            else -> {
                println("Writing a couple more events")
                eventStore.write("some-stream", converter.toCloudEvent(SomeSpecificDomainEvent(currentEventCount + 1)))
                eventStore.write("some-stream", converter.toCloudEvent(SomeSpecificDomainEvent(currentEventCount + 2)))
            }
        }

        subscriptionModel.subscribe("logging-subscription") { e ->
            println("Event number=${converter.toDomainEvent(e).id} received!")
        }

    }
}

fun main(args: Array<String>) {
    runApplication<OccurrentSpringMongoApplication>(*args)
}
```

**Comment by johanhaleby** [comment]
Thank you again! The non-spring-based implementation is not as battle-proven (I've only used Spring Boot in production), but it's very interesting that you get the same problem with the spring implementation. Maybe I've managed to introduce a regression issue somewhere because I currently have Occurrent subscriptions running on atlas in both free- and paid tiers.

However, I'm currently not using the CatchupSubscription in any of my apps. I know it may be a lot to ask, but do you get the same issue if you're not using the "catchup subscription model"? Maybe there's a bug in the `CatchupSubscriptionModel`.

**Comment by johanhaleby** [comment]
Oh, sorry I didn't see that the question was indeed mentioning the `CatchupSubscriptionModel` 🤦‍♂️ But then it would be even more interesting to see if it works without it. My guess/hope is that it will, and if you stop your app, and restart it with the `CatchupSubscriptionModel` enabled again, it'll work. But it definitely sounds like a bug!

**Comment by gavvvr** [comment]
@johanhaleby

> ... do you get the same issue if you're not using the "catchup subscription model"?

> But then it would be even more interesting to see if it works without it.

If I use raw `DurableSubscriptionModel` without wrapping it into `CatchupSubscriptionModel`, then it mostly works as expected and saves the "subscription position".
One a bit unexpected thing which I noticed is that if you do `eventStore.write()` right after `wrappedSubscriptionModel.subscribe` returns, then the 1st event will be missed. But if you add `Thread.sleep(1000)` between `wrappedSubscriptionModel.subscribe` and `eventStore.write()`, all events gets captured by subscription.

> My guess/hope is that it will, and if you stop your app, and restart it with the CatchupSubscriptionModel enabled again, it'll work

Yes, if position was persisted by raw `DurableSubscriptionModel` previously, the `CatchupSubscriptionModel` will continue where the `DurableSubscriptionModel` left off.

**Comment by gavvvr** [comment]
> Is this also the case even if you do:
> 
> `wrappedSubscriptionModel.subscribe(..).waitUntilStarted()`?

@johanhaleby the `waitUntilStarted()` worked out! Did not know about this method :)

So, the `CatchupSubscriptionModel` is buggy for me in both cases (Spring and native) and the raw `DurableSubscriptionModel` works as expected.

**Comment by johanhaleby** [comment]
> > Is this also the case even if you do:
> > `wrappedSubscriptionModel.subscribe(..).waitUntilStarted()`?
> 
> @johanhaleby the `waitUntilStarted()` worked out! Did not know about this method :)

Glad to hear that it works. `waitUntilStarted()` is easy to miss I suppose, I see that it's not well documented besides the javadoc. 

> 
> So, the `CatchupSubscriptionModel` is buggy for me in both cases (Spring and native) and the raw `DurableSubscriptionModel` works as expected.

So `CatchupSubscriptionModel` is the culprit! I'll try to fix the bug when I get some time, hopefully on friday.

**Comment by johanhaleby** [comment]
@gavvvr I think I've fixed the issue now in the main branch. I have some other things I want to change/fix as well, but then I'll make a new release, hopefully during the day.

**Comment by johanhaleby** [comment]
It's been released now, version 0.16.8
